From 50a0a821175836f30c4a439fb44aeed4686b22e6 Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Tue, 23 Apr 2024 09:19:25 +0800
Subject: [PATCH 011/146] mtk: mt76: mt7996: enable ibf capability for mt7992

For the specific sku of mt7992, it supports both ibf and ebf
functionality. The firmware algorithm may decide which type is better
according to the station's beamform capability.

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
---
 mt7996/mcu.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 980bddce..d0eae9f0 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -1663,6 +1663,8 @@ static void
 mt7996_mcu_sta_bfer_tlv(struct mt7996_dev *dev, struct sk_buff *skb,
 			struct ieee80211_vif *vif, struct ieee80211_sta *sta)
 {
+#define EBF_MODE	BIT(0)
+#define IBF_MODE	BIT(1)
 	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
 	struct mt7996_phy *phy = mvif->phy;
 	int tx_ant = hweight16(phy->mt76->chainmask) - 1;
@@ -1701,7 +1703,10 @@ mt7996_mcu_sta_bfer_tlv(struct mt7996_dev *dev, struct sk_buff *skb,
 	else
 		return;
 
-	bf->bf_cap = ebf ? ebf : dev->ibf << 1;
+	bf->bf_cap = ebf ? EBF_MODE : (dev->ibf ? IBF_MODE : 0);
+	if (is_mt7992(&dev->mt76) &&
+	    tx_ant == hweight8(phy->mt76->hw->wiphy->available_antennas_tx))
+		bf->bf_cap |= IBF_MODE;
 	bf->bw = sta->deflink.bandwidth;
 	bf->ibf_dbw = sta->deflink.bandwidth;
 	bf->ibf_nrow = tx_ant;
-- 
2.45.2

